### PR TITLE
Add typing

### PR DESCRIPTION
**Pull request by Rafiot**
As of right now, it is totally useless, I just want to have a draft PR going to keep track of the merge conflicts.

### CODE REVIEW

File: followthemoney/types/registry.py [hunk 1]
```diff
@@ -8,43 +10,40 @@ class Registry(object):
...
         return {t.group: t for t in self.types if t.group is not None}
 
-    def get(self, name):
+    def get(self, name: Union[str, PropertyType]) -> PropertyType:
         """For a given property type name, get its handling object."""
         # Allow transparent re-checking.
         if isinstance(name, PropertyType):
             return name
-        return self._types.get(name)
+        return self._types[name]
```
Comment (pudo) [inline followthemoney/types/registry.py:42]: This is behaviour change. Can this be a separate PR?

File: followthemoney/compare.py [hunk 2]
```diff
@@ -26,22 +31,22 @@
 }
 
 
-def compare(model, left, right):
+def compare(model: Model, left: Mapping, right: Mapping) -> float:
```
Comment (pudo) [inline followthemoney/compare.py:34]: ... Union[Mapping,EntityProxy]

File: followthemoney/export/csv.py [hunk 3]
```diff
@@ -1,26 +1,38 @@
...
-    def _open_csv_file(self, name):
+    def _open_csv_file(self, name) -> Tuple[TextIO, Any]:
         self.directory.mkdir(parents=True, exist_ok=True)
         file_path = self.directory.joinpath('{0}.csv'.format(name))
         handle = open(file_path, mode='w')
         writer = csv.writer(handle, dialect=self.dialect)
         return handle, writer
 
-    def _get_writer(self, schema):
+    def _write_header(self, writer, schema: Schema):
```
Comment (pudo) [inline followthemoney/export/csv.py:27]: Does this apply 1:1 for the Neo4J Bulk exporter, the other user of the CSVMixIn?

File: followthemoney/graph.py [hunk 4]
```diff
@@ -11,26 +16,31 @@ class Node(object):
...
-        self.type = type_
-        self.value = value
-        self.id = type_.node_id_safe(value)
-        self.proxy = proxy
-        self.schema = schema if proxy is None else proxy.schema
+    def __init__(self, type_: PropertyType, value: str,
+                 proxy: Optional[EntityProxy]=None,
+                 schema: Optional[Schema]=None):
+        self.type: PropertyType = type_
+        self.value: Optional[str] = value
```
Comment (pudo) [inline followthemoney/graph.py:23]: Is value optional? That would then be when we get into the whole optional node ID mess.

File: followthemoney/graph.py [hunk 5]
```diff
@@ -11,26 +16,31 @@ class Node(object):
...
+        self.schema: Optional[Schema] = schema if proxy is None else proxy.schema
 
     @property
-    def is_entity(self):
+    def is_entity(self) -> bool:
         return self.type == registry.entity
 
     @property
-    def caption(self):
+    def caption(self) -> Optional[str]:
```
Comment (pudo) [inline followthemoney/graph.py:36]: See: is value optional?

File: followthemoney/graph.py [hunk 6]
```diff
@@ -39,7 +49,9 @@ def to_dict(self):
         }
 
     @classmethod
-    def from_proxy(cls, proxy):
+    def from_proxy(cls, proxy: EntityProxy) -> 'Node':
+        if proxy.id is None:
+            raise Exception('<FIXME>')
```
Comment (pudo) [inline followthemoney/graph.py:54]: FIXME :)

File: followthemoney/graph.py [hunk 7]
```diff
@@ -60,50 +72,59 @@ class Edge(object):
...
             self.schema = proxy.schema
 
     @property
-    def source(self):
+    def source(self) -> Optional[Node]:
         return self.graph.nodes.get(self.source_id)
 
     @property
-    def source_prop(self):
+    def source_prop(self) -> Optional[Property]:
```
Comment (pudo) [inline followthemoney/graph.py:97]: Edge must have a source prop, if you want we can check if it's a valid edge before.

File: followthemoney/graph.py [hunk 8]
```diff
@@ -60,50 +72,59 @@ class Edge(object):
...
+        self.schema: Optional[Schema] = None
         if prop is not None:
             self.weight = prop.specificity(value)
         if proxy is not None:
             self.id = f"{source.id}<{proxy.id}>{target.id}"
             self.schema = proxy.schema
 
     @property
-    def source(self):
+    def source(self) -> Optional[Node]:
```
Comment (pudo) [inline followthemoney/graph.py:93]: Not really optional.

File: followthemoney/graph.py [hunk 9]
```diff
@@ -60,50 +72,59 @@ class Edge(object):
...
         return self.prop
 
     @property
-    def target(self):
+    def target(self) -> Optional[Node]:
         return self.graph.nodes.get(self.target_id)
 
     @property
-    def target_prop(self):
+    def target_prop(self) -> Optional[Property]:
```
Comment (pudo) [inline followthemoney/graph.py:108]: This, however, is optional.

File: followthemoney/graph.py [hunk 10]
```diff
@@ -129,39 +150,41 @@ class Graph(object):
...
         if node.id not in self.nodes:
             self.nodes[node.id] = node
         return self.nodes[node.id]
 
-    def _add_edge(self, proxy, source, target):
-        source = self._get_node_stub(proxy.schema.source_prop, source)
-        target = self._get_node_stub(proxy.schema.target_prop, target)
-        edge = Edge(self, source, target, proxy=proxy)
+    def _add_edge(self, proxy: EntityProxy, source: str, target: str):
+        if proxy.schema.source_prop is None or proxy.schema.target_prop is None:
```
Comment (pudo) [inline followthemoney/graph.py:180]: I don't think either of these checks are needed given how it's invoked.

File: followthemoney/helpers.py [hunk 11]
```diff
@@ -61,16 +66,17 @@ def entity_filename(proxy, base_name=None, extension=None):
...
     names aren't special and may not always be the caption."""
     if entity.schema.is_a('Thing'):
         names = entity.get('name')
         if len(names) > 1:
-            name = registry.name.pick(names)
-            if name in names:
-                names.remove(name)
-            entity.set('name', name)
-            entity.add('alias', names)
+            if isinstance(registry.name, NameType):
```
Comment (pudo) [inline followthemoney/helpers.py:76]: nice catch!

File: followthemoney/model.py [hunk 12]
```diff
@@ -11,11 +13,11 @@
...
-    def __init__(self, path):
-        self.path = path
-        self.schemata = {}
-        self.properties = set()
-        self.qnames = {}
+    def __init__(self, path: str):
+        self.path: str = path
+        self.schemata: Dict[str, Schema] = {}
+        self.properties: Dict[str, Property] = {}
+        self.qnames: Dict[str, str] = {}
```
Comment (pudo) [inline followthemoney/model.py:20]: is qnames str not property?

File: followthemoney/model.py [hunk 13]
```diff
@@ -62,49 +64,59 @@ def get_type_schemata(self, type_):
...
         """
-        left = self.get(left) or self.get(right)
-        right = self.get(right) or self.get(left)
-        if left.is_a(right):
-            return left
-        if right.is_a(left):
-            return right
-        common = list(left.schemata.intersection(right.schemata))
+        left_schema = self.get(left) or self.get(right)
+        if left_schema is None:
```
Comment (pudo) [inline followthemoney/model.py:87]: Behaviour change, the previous was to return the opposing side's schema, not throw. Can we make this a separate PR?

File: followthemoney/model.py [hunk 14]
```diff
@@ -62,49 +64,59 @@ def get_type_schemata(self, type_):
...
             if schema.is_a(specific):
                 specific = schema
         return specific
 
-    def make_entity(self, schema, key_prefix=None):
+    def make_entity(self, schema: Mapping, key_prefix=None) -> EntityProxy:
         return EntityProxy(self, {'schema': schema}, key_prefix=key_prefix)
 
-    def get_proxy(self, data, cleaned=True):
+    def get_proxy(self, data: Mapping, cleaned=True) -> EntityProxy:
```
Comment (pudo) [inline followthemoney/model.py:116]: Union[Mapping, EntityProxy]

File: followthemoney/model.py [hunk 15]
```diff
@@ -62,49 +64,59 @@ def get_type_schemata(self, type_):
...
+            raise InvalidData(msg % (left_schema, right_schema))
 
         specific = common[0]
         for schema in common[1:]:
             if schema.is_a(specific):
                 specific = schema
         return specific
 
-    def make_entity(self, schema, key_prefix=None):
+    def make_entity(self, schema: Mapping, key_prefix=None) -> EntityProxy:
```
Comment (pudo) [inline followthemoney/model.py:113]: Union[str, Schema], no?

File: followthemoney/namespace.py [hunk 16]
```diff
@@ -69,14 +77,19 @@ def apply(self, proxy):
             if prop.type != registry.entity:
                 continue
             for value in signed.pop(prop):
-                value = get_entity_id(value)
-                signed.add(prop, self.sign(value))
+                _value = get_entity_id(value)
+                signature = self.sign(_value)
+                if signature:
+                    signed.add(prop, signature)
+                else:
```
Comment (pudo) [inline followthemoney/namespace.py:84]: signed.add handles None transparently, no need to bother with it here.

File: followthemoney/property.py [hunk 17]
```diff
@@ -1,36 +1,36 @@
...
-        self.stub = data.get('stub', False)
+        self._label: str = data.get('label', name)
+        self._description: Optional[str] = data.get('description')
+        self.hidden: bool = data.get('hidden', False)
+        self.stub: bool = data.get('stub', False)
 
-        type_ = data.get('type', 'string')
-        self.type = registry.get(type_)
-        if self.type is None:
-            raise InvalidModel("Invalid type: %s" % type_)
```
Comment (pudo) [inline followthemoney/property.py:31]: See registry, weird behaviour change.

File: followthemoney/proxy.py [hunk 18]
```diff
@@ -276,7 +292,7 @@ def __eq__(self, other):
         return self.id == other.id
 
     @classmethod
-    def from_dict(cls, model, data, cleaned=True):
+    def from_dict(cls, model: Model, data, cleaned: bool=True):
```
Comment (pudo) [inline followthemoney/proxy.py:295]: data: Union[Mapping, EntityProxy]

File: followthemoney/types/registry.py (hunk 1 above)
Comment (Rafiot) [reply followthemoney/types/registry.py:42]: The reason I changed it is because using `get` turns the return type to `Optional[PropertyType]`... and everything that calls `get` or `registry[<name>]` can be `None`.

So I can change that, but that will also mean a lot of type checking down the line:

```
$ (git::typing) mypy . 
followthemoney/types/registry.py:46: error: List comprehension has incompatible type List[Optional[PropertyType]]; expected List[PropertyType]
followthemoney/schema.py:117: error: Item "None" of "Optional[PropertyType]" has no attribute "name"
followthemoney/proxy.py:186: error: Argument 1 to "get_type_values" of "EntityProxy" has incompatible type "Optional[PropertyType]"; expected "PropertyType"
followthemoney/proxy.py:202: error: Item "None" of "Optional[PropertyType]" has no attribute "rdf"
followthemoney/proxy.py:224: error: Argument 1 to "get_type_values" of "EntityProxy" has incompatible type "Optional[PropertyType]"; expected "PropertyType"
followthemoney/proxy.py:228: error: Argument 1 to "get_type_values" of "EntityProxy" has incompatible type "Optional[PropertyType]"; expected "PropertyType"
followthemoney/property.py:32: error: Incompatible types in assignment (expression has type "Optional[PropertyType]", variable has type "PropertyType")
followthemoney/namespace.py:24: error: Item "None" of "Optional[PropertyType]" has no attribute "clean"
followthemoney/graph.py:55: error: Argument 1 to "Node" has incompatible type "Optional[PropertyType]"; expected "PropertyType"
followthemoney/compare.py:20: error: Dict entry 0 has incompatible type "Optional[PropertyType]": "int"; expected "PropertyType": "float"
followthemoney/compare.py:21: error: Dict entry 1 has incompatible type "Optional[PropertyType]": "int"; expected "PropertyType": "float"
followthemoney/compare.py:22: error: Dict entry 2 has incompatible type "Optional[PropertyType]": "float"; expected "PropertyType": "float"
followthemoney/compare.py:23: error: Dict entry 3 has incompatible type "Optional[PropertyType]": "float"; expected "PropertyType": "float"
followthemoney/compare.py:24: error: Dict entry 4 has incompatible type "Optional[PropertyType]": "float"; expected "PropertyType": "float"
followthemoney/compare.py:25: error: Dict entry 5 has incompatible type "Optional[PropertyType]": "float"; expected "PropertyType": "float"
followthemoney/compare.py:26: error: Dict entry 6 has incompatible type "Optional[PropertyType]": "float"; expected "PropertyType": "float"
followthemoney/compare.py:27: error: Dict entry 7 has incompatible type "Optional[PropertyType]": "float"; expected "PropertyType": "float"
followthemoney/compare.py:28: error: Dict entry 8 has incompatible type "Optional[PropertyType]": "float"; expected "PropertyType": "float"
followthemoney/compare.py:29: error: Dict entry 9 has incompatible type "Optional[PropertyType]": "float"; expected "PropertyType": "float"
followthemoney/compare.py:30: error: Dict entry 10 has incompatible type "Optional[PropertyType]": "float"; expected "PropertyType": "float"
followthemoney/export/graph.py:9: error: Item "None" of "Optional[PropertyType]" has no attribute "name"
Found 21 errors in 8 files (checked 101 source files)
```

File: followthemoney/compare.py (hunk 2 above)
Comment (Rafiot) [reply followthemoney/compare.py:34]: Would it make sense to make `EntityProxy` inherit on `Mapping` then? because right now, it is not possible to call `dict(EntityProxy)`:
https://github.com/alephdata/followthemoney/blob/91e56b450a9c1bc29221e40cfc2512dfd1f97dfd/followthemoney/proxy.py#L25

File: followthemoney/export/csv.py (hunk 3 above)
Comment (Rafiot) [reply followthemoney/export/csv.py:27]: .... I realized that I missed adding typing in the neo4j exporter, so I'm pretty sure it doesn't vaguely work. I'll look at that.

File: followthemoney/graph.py (hunk 4 above)
Comment (Rafiot) [reply followthemoney/graph.py:23]: It doesn't have to, but then `EntityType.caption` has to return a string, and not `None`: https://github.com/alephdata/followthemoney/blob/91e56b450a9c1bc29221e40cfc2512dfd1f97dfd/followthemoney/types/entity.py#L32

File: followthemoney/graph.py (hunk 6 above)
Comment (Rafiot) [reply followthemoney/graph.py:54]: Yeah, I wasn't sure which kind of exception I should be using ;)

File: followthemoney/graph.py (hunk 8 above)
Comment (Rafiot) [reply followthemoney/graph.py:93]: So should I do `self.graph.nodes.get[self.source_id]` instead?

File: followthemoney/graph.py (hunk 7 above)
Comment (Rafiot) [reply followthemoney/graph.py:97]: I should do the following then:
1. `self.schema.source_prop.reverse` cannot be `None` (given `self.schema.source_prop` exists)
2. `self.prop` cannot be `None`

Comment (Rafiot) [reply followthemoney/graph.py:97]: Okay, I missed a thing. In order to initialize an edge, we need to have at least a `prop` OR a `proxy` (to initialize the schema)

File: followthemoney/helpers.py (hunk 11 above)
Comment (Rafiot) [reply followthemoney/helpers.py:76]: Thanks ;) mypy is pretty good at it

File: followthemoney/graph.py (hunk 10 above)
Comment (Rafiot) [reply followthemoney/graph.py:180]: I can add a type ignore on both calls to `self._get_node_stub`, but as `proxy.schema.source_prop` and `proxy.schema.target_prop` can be `None`, mypy will not like it if I don't do the test first.

File: followthemoney/model.py (hunk 12 above)
Comment (Rafiot) [reply followthemoney/model.py:20]: woops, good catch... I changed the type of `self.properties` and `generate` is broken right now.

File: followthemoney/model.py (hunk 13 above)
Comment (Rafiot) [reply followthemoney/model.py:87]: Sure, but the current code fails at runtime if either `left` or `right` are `None`: 
https://github.com/alephdata/followthemoney/blob/91e56b450a9c1bc29221e40cfc2512dfd1f97dfd/followthemoney/model.py#L86
https://github.com/alephdata/followthemoney/blob/91e56b450a9c1bc29221e40cfc2512dfd1f97dfd/followthemoney/model.py#L88

File: followthemoney/model.py (hunk 14 above)
Comment (Rafiot) [reply followthemoney/model.py:116]: For that, `EntityProxy` need to inherit `Mapping`

File: followthemoney/model.py (hunk 15 above)
Comment (Rafiot) [reply followthemoney/model.py:113]: woops, yes

File: followthemoney/namespace.py (hunk 16 above)
Comment (Rafiot) [reply followthemoney/namespace.py:84]: :+1:

File: followthemoney/property.py (hunk 17 above)
Comment (Rafiot) [reply followthemoney/property.py:31]: Yeah, I did that because get now always returns a `PropertyType`, or fails. See comment over there.

File: followthemoney/proxy.py (hunk 18 above)
Comment (Rafiot) [reply followthemoney/proxy.py:295]: Same, we can do that if `EntityProxy` inherits  `Mapping`

File: followthemoney/compare.py (hunk 2 above)
Comment (Rafiot) [reply followthemoney/compare.py:34]: ... or at least I'm super confused how that is supposed to work right now.

Based on the doc, I cannot see how you can pass an `EntityProxy` to `dict` -> https://docs.python.org/3.8/library/stdtypes.html#dict

File: followthemoney/export/csv.py (hunk 3 above)
Comment (Rafiot) [reply followthemoney/export/csv.py:27]: So, I checked, yes, it will works. the method `_write_header` moved from `CSVExporter` to `CSVMixin` is overwritten in `Neo4JCSVExporter`

**Comment by Rafiot** [comment]
Alright, I started to play with the code... and we (obviously) have circular imports all over the place. I'll figure something out.

**Comment by Rafiot** [comment]
(rebased to HEAD)

**Comment by Rafiot** [comment]
The test cases are passing again :)

**Comment by pudo** [comment]
Wow, that's a lot of work you've done on this already. A bit more complex than I thought, I'll admit :) 

I've been doing a bit of work on making the whole thing pass a linter on this branch here: [origin/typing](https://github.com/alephdata/followthemoney/tree/typing).

The circular imports issue is really weird and leaves me very uncomfortable, would it be helpful is we broke compatibility below 3.7 (I hear theres a `__future__` that might be helpful)?

Another thing that I cannot quite wrap my head around yet is the method signatures on the type system - if I break up a method signature into two lines, `mypy` doesn't seem to accept the `type: ignore` any more, and breaks on differences between the upstream signature and the local one. Does this mean we can either have a happy linter or a happy type check or is there some workaround to this?

**Comment by Rafiot** [comment]
Breaking compatibility below 3.7 is totally cool with me. I cannot force that before the end of this year on (Py)MISP side, because my users would kill me, but FTM will never be a required dependency, so that's definitely not a problem.

I think (?) all the circular imports are solved by using the `TYPE_CHECKING` trick. If I understand the documentation right, relying on `from __future__ import annotations` will only make the typing slightly nicer by avoiding using the quotes for classes defined in the same file: https://mypy.readthedocs.io/en/stable/common_issues.html#import-cycles
I'm not a fan of the `TYPE_CHECKING` trick, I find it pretty dirty, but for what I understand that's the only way to do that so mypy doesn't fail when we have different files needing definitions from each other.

And for the linter vs typing, this is the only trick I found: https://mypy.readthedocs.io/en/stable/common_issues.html#silencing-linters 
But I find it pretty dirty, unpleasant to read, and it doesn't seem to work on multiple lines: https://github.com/python/mypy/issues/1032 (this is absolutely disgusting and my linter absolutely hates that, but mypy is happy... But it seems to be supported properly with Python 3.8)

So for what I see right now, the remaining things that are failing with mypy are: 
1. incompatible with supertype (x2) -> linter vs. mypy 
2. `"EntityProxy" has incompatible type "Union[EntityProxy, Mapping[Any, Any]]"; expected "Mapping[Any, Any]"` -> I still don't understand how it is possible to have a `data` of type `EntityProxy` passed to a `dict()` (https://github.com/alephdata/followthemoney/blob/typing/followthemoney/proxy.py#L37) In my books, it only works if `EntityProxy` is an `Iterator`
3. `_get_node_stub" of "Graph" has incompatible type "Optional[Property]"; expected "Property"` because `Schema.source_prop` and `Schema.target_prop` can return `None`.

So yeah, looks good to me!

**Comment by Rafiot** [comment]
Re. circular imports: https://github.com/alephdata/followthemoney/pull/246

**Comment by Rafiot** [comment]
Part of https://github.com/alephdata/followthemoney/tree/typing
